### ISSUE TITLE
`jnp.nonzero` zero padding behavior is confusing

### ISSUE THREAD

**Issue by thisiscam** [google/jax#7129]
I noticed that `jax.numpy.nonzero` now allows the wonderful `size` parameter which allows jitting: #6501. This is wonderful! Thanks!

However, the semantics is that missing zeros are zero-padded. This seems unworkable in some situations, consider:
```python
x = jnp.zeros(5)
print(jnp.nonzero(x, size=1))   # Prints [ 0 ]

x = jnp.ones(5)
print(jnp.nonzero(x, size=1))   # Prints [ 0 ]
```

There's no way to tell between these two cases.

Instead, perhaps you could consider something like `-1`-padding? 
Though, performance-wise, I'm not sure how hard it would be to make that on-par with the current implementation

**Comment by thisiscam** [comment]
Note that this only occurs when `size=1`. 
 
So perhaps at least we should document this corner case since this seems like a problem that could cause unexpected program behaviors.

**Comment by jakevdp** [comment]
Thanks for the question – we chose to make the padding zero because it's the natural value that falls out of the algorithm. Zero is a natural value in some applications, and any other fill value would require additional post-processing of the result. If you want to configure the fill value, you can do so with a short utility routine. For example:
```python
import jax.numpy as jnp

def nonzero_with_configurable_padding(arr, *, size=None, fill_value=-1):
  out = jnp.nonzero(arr, size=size)
  if size is not None:
    num_nonzero = (arr != 0).sum()
    mask = jnp.arange(len(out[0])) < num_nonzero
    out = tuple(jnp.where(mask, entry, fill_value) for entry in out)
  return out

dx = jnp.zeros(5)
print(nonzero_with_configurable_padding(x, size=1))
# [ -1 ]

x = jnp.ones(5)
print(nonzero_with_configurable_padding(x, size=1))
# [ 0 ]
```

**Comment by thisiscam** [comment]
@jakevdp Thanks for the explanation and your example. 
A couple questions for this overall issue:
1) Is there an alternative algorithm that can a) use arbitrary `fill_value` b) retain the same efficiency?
2) Should we document this behavior for `size=1`? I would say that we could even ban `size=1` since it produces ambiguous return value.
3) Perhaps you can add this implementation of `fill_value` parameter to `nonzero`? The default is zero so the fast code path is enabled. 

Ideally, someone will think of solution to 1), but before that, I think we should do either 2) or 3). 
WDYT?

**Comment by jakevdp** [comment]
I think the behavior for size=1 is already documented, no? The true size is zero, so the result is padded with zeros up to the requested size.

**Comment by thisiscam** [comment]
> The true size is zero, so the result is padded with zeros up to the requested size.

Well, sure. I'm simply thinking that we should add a note, that `size=1` will have this problem of ambiguous output. 
I mean, from a user's perspective, I do hope that when I write some code with `size=1`, I can get fair amount of warning of this behavior. Maybe it's just me, but I only realized this issue after I observed unexpected output. 

I can do a PR on the docs. 
The only reason I haven't done so is because I'm not sure if 1) or 3) is a viable/better option?

**Comment by jakevdp** [comment]
I think a `fill_value` option may be a reasonable feature to add. The reason I didn't before is that it adds a maintenance/testing burden for something that didn't have any concrete use-case to motivate it, and the relevant information (the true number of nonzeros) is straightforward to determine separately.

Can you give an example of code you might write with a configurable fill value?

**Comment by thisiscam** [comment]
Sure. The code I had which ran into this problem was a jittable [Schur-Transform](https://en.wikipedia.org/wiki/Lehmer–Schur_algorithm). The implementation is pretty short so I'd paste it here (note that most of it is just comments). 

```python
import jax.numpy as jnp


def _first_non_zero(x):
  return jnp.nonzero(x, size=1)[0][0]
  # return jnp.nonzero(x, size=1, fill_value=x.size)[0][0]


def schur_transform(coeffs,
                    n_leading_zeros: int):
  """Schur Transform of a polynomial.
  
  Computes 
  ..math::

    T(p) = \\bar{p(0)}p - \\bar{p^*(0)}p^*
  
  where :math:`p^*(z)` is the reciprocal adjoint polynomial 
  ..math::

    p^*(z) = z^n \cdot \\bar{p(z^{-1})}

  and bar denotes complex conjugation.

  See https://en.wikipedia.org/wiki/Lehmer–Schur_algorithm for more details.


  Args:
    coeffs: a complex array of polynomial coefficients, with possibily some 
    leading zero terms.
    n_leading_zeros: the number of leading zeros in the polynomial.

  Returns:
    the transformed polynomial coefficients, and the number of leading zeros in
    it.
  """
  # The reciprocal adjoint polynomial
  coeffs_ra = jnp.roll(coeffs.conj(), shift=-n_leading_zeros, axis=0)
  # Schur Transformation
  coeffs = (coeffs[-1].conj() * coeffs - coeffs_ra[-1].conj() * coeffs_ra)
  # Update number of leading zeros
  n_leading_zeros = _first_non_zero(coeffs)
  return coeffs, n_leading_zeros
```

The test could be
```
coeffs = jnp.array([1, 2, 1], dtype=jnp.complex64) # f(x) = x^2 + 2x + 1
transformed_coeffs, n_leading_zeros = schur_transform(coeffs, n_leading_zeros=0)
assert jnp.all(transformed_coeffs == 0)
assert n_leading_zeros == 2 # Currently fails because `fill_value` is not available.
```
